Contour plot

jonas-kleck opened the pull request:
Added the possibility to specify the confidence level as fraction instead of multiples of sigma for the ContourProfiler. Aligned the x-axes of the contour plots and the profile plots. Added the new features to one of the example scipts.

jonas-kleck committed:
added possibility to specify cl instead of multiples of sigma for the contour plotting. Aligned the x axes of the contour plots and the profile plots

jonas-kleck committed:
added the new functions to the example script

JohannesGaessler commented on kafe2/fit/tools/contours_profiler.py:
```diff
@@ -94,6 +98,33 @@ def format_data_short(self, value):
+        _sign = np.sign(_number_of_sigma)
+        if _sign != 0:
+            _cl = ConfidenceLevel(n_dimensions=1, sigma=_sign*_number_of_sigma)
+            _cl_label = _cl.cl*_sign * 50
```
If I understand this code correctly it will result in confidence levels with a leading minus sign for ticks below the central value. I think that is incorrect since a confidence level is by definition positive and in [0, 1].

jonas-kleck replied:
So it should rather look like this?
<img width="362" height="387" alt="grafik" src="https://github.com/user-attachments/assets/91809343-3dd4-4ea3-a215-07233bd7c823" />

JohannesGaessler commented on kafe2/fit/tools/contours_profiler.py:
```diff
@@ -186,12 +218,17 @@ def __init__(
         :param contour_smoothing_sigma: apply a smoothing Gaussian filter with this sigma parameter to each contour
                                         (default is ``0.0``, meaning no smoothing)
         :type contour_smoothing_sigma: float
+        :param use_as_cl_values: If ``True``, the values from contour_sigma_values are interpreted as confidence level
```
```suggestion
        :param use_as_cl_values: If ``True``, the values from contour_sigma_values are interpreted as confidence levels instead
```

JohannesGaessler commented on kafe2/fit/tools/contours_profiler.py:
```diff
@@ -434,9 +471,10 @@ def plot_profile(
+        label_ticks="sigma",
         label_fit_minimum=True,
         font_scale=1.0,
+        use_fractions=False,
```
I think it would make more sense to instead add an argument for the sigma interval for the ticks, with a default of 1.0.

JohannesGaessler commented on kafe2/fit/tools/contours_profiler.py:
```diff
@@ -571,14 +612,19 @@ def plot_profile(
                     _form_x = ScalarFormatter(sigma=_par_err, n_significant_digits=2)
                     _axes.xaxis.set_major_formatter(_form_x)
+                else:
+                    raise ValueError("Unknown label tick naming convention '%s'! " "Must be one of: ('cl', 'sigma', 'value')" % (label_ticks,))
```
Preferably use f strings in new code (e.g. `f"The value is {variable_name}"`).

JohannesGaessler reviewed (commented):
Preferably add relevant plots to PRs so that we can see what the new plots are supposed to look like.

If possible, please add unit tests. It's not feasible to check whether the plots themselves are correct, they just need to check that the program doesn't crash.

jonas-kleck committed:
reworked the ticks of the cl in the contour plots

jonas-kleck committed:
added unittest for new ContoursProfiler functions

jonas-kleck committed:
added documentation about the new features of ContoursProfiler

jonas-kleck committed:
formatting changes

jonas-kleck commented:
All above comments should be incorporated. A contour plot with 30% and 90% contours should look like this: 
<img width="984" height="977" alt="Screenshot 2025-08-28 203710" src="https://github.com/user-attachments/assets/d1919238-2743-4d99-a0c0-14206503dbe3" />

JohannesGaessler commented:
Sorry, I forgot that for 2d confidence areas you cannot meaningfully label the axes as percentages. As described by Günter, the axes should be labeled in absolute units of the parameter instead.

JohannesGaessler commented on doc/src/parts/user_guide.rst:
```diff
@@ -559,6 +559,22 @@ First, a ContoursProfiler object must be created from a Fit object:
     cpf = ContoursProfiler(fit)
 		
+By default, the one and two :math:`\sigma` contours are created. In case other values are desired,
+they can be specified in the creation of the ContoursProfiler object:
```
```suggestion
By default, the one and two :math:`\sigma` contours are created.
Other values can be specified in the ContoursProfiler constructor:
```

JohannesGaessler reviewed (commented):
I have a suggestion for the `ContoursProfiler` constructor interface. We can set the default for `contour_sigma_values` to `None` and add a new argument `contour_cl_values`, also `None` by default. If both are `None`, use 1/2 sigma. If one is specified, use that one. If both are specified, raise an error. I think that would be a more intuitive interface than the `use_as_cl_values` flag.

jonas-kleck commented:
> Sorry, I forgot that for 2d confidence areas you cannot meaningfully label the axes as percentages. As described by Günter, the axes should be labeled in absolute units of the parameter instead.

So I can completely remove the option to show axis labels as percentages?

GuenterQuast commented:
Hallo Jonas,

die Anzeige in "percentages" ist wirklich schwer zu verstehen und absolut unüblich.
Also in der Tat besser entfernen.

 Viele Grüße,
    Günter


jonas-kleck schrieb:
> *jonas-kleck* left a comment (PhiLFitters/kafe2#256) <https://github.com/PhiLFitters/kafe2/pull/256#issuecomment-3261723641>
> 
>     Sorry, I forgot that for 2d confidence areas you cannot meaningfully label the axes as percentages. As described by
>     Günter, the axes should be labeled in absolute units of the parameter instead.
> 
> So I can completely remove the option to show axis labels as percentages?
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/PhiLFitters/kafe2/pull/256#issuecomment-3261723641>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEIAGSIGN6XWM2A4YAQCUW33RKVMPAVCNFSM6AAAAACEOSUWB2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTENRRG4ZDGNRUGE>.
> You are receiving this because you are subscribed to this thread.Message ID:
> ***@***.***>
> 

-- 
 ______________________________________________
|             Günter Quast                     |
| Kösliner Straße 5 | Tel. +49 (0)721 3529993  |
| 76139 Karlsruhe   | Mobil: +49 1520 9893977  |
|      E-Mail:   ***@***.***       |
|______________________________________________|

jonas-kleck committed:
removed optino to display ticks as percentages, changed interface of contours profiler constructor

jonas-kleck committed:
improved style

jonas-kleck commented:
I removed the option to display tick labels in percentages and implemented @JohannesGaessler suggestion for the interface of the ContoursProfiler constructor. The Contour plot using tick labels in the same units as the parameter: 
<img width="500" height="500" alt="Screenshot 2025-09-08 214908" src="https://github.com/user-attachments/assets/30785fa6-e25c-43c3-aea0-c35b03fe6e31" />

GuenterQuast commented:
Thanks - looks allright. Some remarks/imporvements:

 - the contours should always be fully contained in the graph;
   the plot x/y-limits could be derived from the uncertainies by
   converting Cl to sigma. As an alternative, +/- 3.5 sigma could
   always be used, unless the requested confidence level is very large.

 - as a cross-check of the implementation, could you show that the 39% CL contour
    corresponds to one sigma ?

Thanks, Günter

jonas-kleck schrieb:
> *jonas-kleck* left a comment (PhiLFitters/kafe2#256) <https://github.com/PhiLFitters/kafe2/pull/256#issuecomment-3269462414>
> 
> I removed the option to display tick labels in percentages and implemented @JohannesGaessler
> <https://github.com/JohannesGaessler> suggestion for the interface of the ContoursProfiler constructor. The Contour plot
> using tick labels in the same units as the parameter:
> Screenshot.2025-09-08.214908.png (view on web)
> <https://github.com/user-attachments/assets/30785fa6-e25c-43c3-aea0-c35b03fe6e31>
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/PhiLFitters/kafe2/pull/256#issuecomment-3269462414>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEIAGSLLLRTP6KPQI2OKVMT3R2EGNAVCNFSM6AAAAACEOSUWB2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTENRZGQ3DENBRGQ>.
> You are receiving this because you commented.Message ID: ***@***.***>
> 

-- 
 ______________________________________________
|             Günter Quast                     |
| Kösliner Straße 5 | Tel. +49 (0)721 3529993  |
| 76139 Karlsruhe   | Mobil: +49 1520 9893977  |
|      E-Mail:   ***@***.***       |
|______________________________________________|

jonas-kleck committed:
Changed x and y limits for the contour plots

jonas-kleck commented:
The plot limits in x and y direction for the contour Plots is now by default +- 3.5 sigma. If the contour exceeds these limits, they are changed, so that the whole contour should be contained.

GuenterQuast commented:
Ran the code and looks ok - the number of ticks labes is still large - might be reduced from
7 to 5 or 4.

jonas-kleck schrieb:
> *jonas-kleck* left a comment (PhiLFitters/kafe2#256) <https://github.com/PhiLFitters/kafe2/pull/256#issuecomment-3284800273>
> 
> The plot limits in x and y direction for the contour Plots is now by default +- 3.5 sigma. If the contour exceeds these
> limits, they are changed, so that the whole contour should be contained.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/PhiLFitters/kafe2/pull/256#issuecomment-3284800273>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEIAGSLWVBUFRVCESMGZK3T3SKRABAVCNFSM6AAAAACEOSUWB2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTEOBUHAYDAMRXGM>.
> You are receiving this because you commented.Message ID: ***@***.***>
> 

-- 
 ______________________________________________
|             Günter Quast                     |
| Kösliner Straße 5 | Tel. +49 (0)721 3529993  |
| 76139 Karlsruhe   | Mobil: +49 1520 9893977  |
|      E-Mail:   ***@***.***       |
|______________________________________________|

jonas-kleck committed:
changed default plot limits for contour profiler

jonas-kleck commented:
Now there should be only 5 sigma ticks as default

GuenterQuast commented:
Results looks good - ok for me to merge. Any further comments on code 
style or quality ?

Am 17.09.25 um 11:32 schrieb jonas-kleck:
> *jonas-kleck* left a comment (PhiLFitters/kafe2#256) 
> <https://github.com/PhiLFitters/kafe2/pull/256#issuecomment-3302138106>
>
> Now there should be only 5 sigma ticks as default
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/PhiLFitters/kafe2/pull/256#issuecomment-3302138106>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AEIAGSIVNKCSYON5K64JG433TETBTAVCNFSM6AAAAACEOSUWB2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTGMBSGEZTQMJQGY>.
> You are receiving this because you commented.Message ID: 
> ***@***.***>
>

JohannesGaessler commented on doc/src/parts/user_guide.rst:
```diff
@@ -559,6 +559,23 @@ First, a ContoursProfiler object must be created from a Fit object:
     from kafe2 import ContoursProfiler
     cpf = ContoursProfiler(fit)
 		
+By default, the one and two :math:`\sigma` contours are created. Other values can be
```
```suggestion
By default, the one and two :math:`\sigma` contours are calculated. Other values can be
```

JohannesGaessler commented on doc/src/parts/user_guide.rst:
```diff
@@ -559,6 +559,23 @@ First, a ContoursProfiler object must be created from a Fit object:
+.. code-block:: python
+
+    from kafe2 import ContoursProfiler
+    cpf = ContoursProfiler(fit, contour_sigma_values = (1,2,3))
```
```suggestion
    cpf = ContoursProfiler(fit, contour_sigma_values=(1, 2, 3))
```

This is I think the I think officially recommended code style.

JohannesGaessler commented on doc/src/parts/user_guide.rst:
```diff
@@ -559,6 +559,23 @@ First, a ContoursProfiler object must be created from a Fit object:
+.. code-block:: python
+
+    from kafe2 import ContoursProfiler
+    cpf = ContoursProfiler(fit, contour_cl_values = (0.9, 0.95))
```
```suggestion
    cpf = ContoursProfiler(fit, contour_cl_values=(0.9, 0.95))
```

JohannesGaessler commented on doc/src/parts/user_guide.rst:
```diff
@@ -583,7 +600,7 @@ or to display only selected confidence contours:
 Consult the api documentation on details how to restrict the displayed ranges
 or for special plot options.
-
+ 
```
```suggestion

```

JohannesGaessler commented on examples/003_profiling/02_double_slit.py:
```diff
@@ -103,16 +104,17 @@ def intensity(theta, I_0, b, g, varlambda):
 fit.do_fit()
 
-cpf = ContoursProfiler(fit)
+cpf = ContoursProfiler(fit,contour_sigma_values = None, contour_cl_values = (0.5,0.6))
```
```suggestion
cpf = ContoursProfiler(fit, contour_cl_values=(0.5, 0.6))
```

50% and 60% seem like strange confidence levels to me. Is there a reason why these specific values were chosen?

jonas-kleck replied:
No I just played around with how the plot ranges behave when I use smaller or larger contours. I can also set them to other values. Like 90% and 95%?

JohannesGaessler replied:
If there are no issues with 90% and 95% I think those would be preferable as they are more typical.

JohannesGaessler commented on kafe2/fit/tools/contours_profiler.py:
```diff
@@ -163,9 +164,10 @@ def __init__(
         profile_bound=2.45,
         contour_points=100,
-        contour_sigma_values=(1.0, 2.0),
+        contour_sigma_values=None,
```
```suggestion
        contour_sigma_values=None,
        contour_cl_values=None,
```

I think the two arguments should be next to each other in the signature. In principle this would lead to breakage if someone is calling this constructor with only positional arguments but I don't expect that we have even a single user doing this.

JohannesGaessler commented on kafe2/fit/tools/contours_profiler.py:
```diff
@@ -186,11 +188,29 @@ def __init__(
                                         (default is ``0.0``, meaning no smoothing)
         :type contour_smoothing_sigma: float
+        :param use_as_cl_values: If ``True``, the values from contour_sigma_values are interpreted as confidence level instead
+        :type use_as_cl_values: bool
```
This docstring needs to be adjusted.

jonas-kleck committed:
edited contour plot doc and example

jonas-kleck committed:
adjusted code style

JohannesGaessler commented on kafe2/fit/tools/contours_profiler.py:
```diff
@@ -183,14 +185,32 @@ def __init__(
         :type contour_points: int
         :param contour_sigma_values: evaluate and show contours for these confidences (in sigma)
         :type contour_sigma_values: iterable of float
+        :param contour_cl_values: evaluate and show contours for these confidences (as probability)
+        :type contour_cl_values: iterable or float
```
```suggestion
        :param contour_sigma_values: evaluate and show contours for these confidence levels (specified as sigma)
        :type contour_sigma_values: iterable of float
        :param contour_cl_values: evaluate and show contours for these confidence levels (specified as probability)
        :type contour_cl_values: iterable or float
```

jonas-kleck committed:
edited contours_profiler docstring

Co-authored-by: Johannes Gäßler <johannesg@5d6.de>

JohannesGaessler commented:
Thank you.
